### PR TITLE
[WIP] Implement Accumulator Framework and Compound Accumulator for aggregation

### PR DESCRIPTION
**Pull request by preethiraghavan1**
## Description
Implementing an abstract class for Accumulator and CompoundAccumulator class. Have not implemented serialization yet. Wanted to share this to see if this is in the right direction.

## How has this been tested?
- Unit tests

## Checklist
- [x ] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [ x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [ ] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [ x] My changes are covered by tests

### CODE REVIEW

File: pipeline_dp/accumulator.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+
+class Accumulator(abc.ABC):
+  """
+  Performs aggregations
+  """
+
+  @classmethod
+  def merge(cls, accumulators: typing.Iterable[
+    'Accumulator'] = None) -> 'Accumulator':
```
Comment (preethiraghavan1) [inline pipeline_dp/accumulator.py:12]: How do you like the idea of args type argument here.
```
merge(cls, *accumulators: 'Accumulator' = None) 
```

A list need not be created when a single accumulator is involved.
But when a list is involved, then they have to unpack it merge(*list_of_acc)

May not be super helpful, but wanted to know your thoughts on this.

File: pipeline_dp/accumulator.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+  def add_value(self, v):
+    # adds the value to each accumulator
+    for accumulator in self.accumulators:
+      accumulator.add_value(v)
+    return self
+
+  def add_accumulator(self, accumulator: 'CompoundAccumulator') -> \
+    'CompoundAccumulator':
+    # merges the accumulators of the CompoundAccumulators
+    self.accumulators.extend(accumulator.accumulators)
```
Comment (dvadym) [inline pipeline_dp/accumulator.py:77]: Sry, I didn't explained that, the idea that we merge CompoundAcculators which contains accumulators of the same type and in the same order.
**add_accumulator()** will be called when we will have 1 accumulators, which have results from different worker machines.

E.g. let's assume that
```
compound_accumulator1.accumulators = [count_accumulator1, sum_accumulator1]
compound_accumulator2.accumulators = [count_accumulator2, sum_accumulator2]
```
then `compound_accumulator1.add_accumulator(compound_accumulator2) `should do
```
count_accumulator1.add_accumulator(count_accumulator2)
sum_accumulator1.add_accumulator(sum_accumulator2)
```

i.e. `add_accumulator(acc)` accumulates data from `acc`.

File: tests/accumulator_test.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+    compound_accumulator.add_accumulator(to_be_added_compound_accumulator)
+    to_be_added_compound_accumulator.add_value(5)
+
+    computed_metrics = compound_accumulator.compute_metrics()
+    self.assertEqual(len(computed_metrics), 2)
+    self.assertEqual(computed_metrics, [4.5, 41])
+
+
+class MeanAccumulator(pipeline_dp.Accumulator):
```
Comment (dvadym) [inline tests/accumulator_test.py:61]: I like idea with Mean and SumOfSquares accumulators for testing!

**Review by dvadym — COMMENTED** [review]
Thanks it looks great! I left some comments

File: pipeline_dp/accumulator.py (hunk 1 above)
Comment (preethiraghavan1) [reply pipeline_dp/accumulator.py:14]: Got it! Thanks!

File: pipeline_dp/accumulator.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
+import abc
+import typing
+
+
+class Accumulator(abc.ABC):
+  """
+  Performs aggregations
```
[the remark this answers is not in the document]
Comment (preethiraghavan1) [reply pipeline_dp/accumulator.py]: Done

File: pipeline_dp/accumulator.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    Merges the accumulators and creates an Accumulator
+    Args:
+      accumulators:
+
+    Returns: Accumulator instance with merged values
+    """
+    return cls(accumulators)
+
+  @abc.abstractmethod
+  def add_value(self, v):
```
[the remark this answers is not in the document]
Comment (preethiraghavan1) [reply pipeline_dp/accumulator.py]: Done

File: pipeline_dp/accumulator.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    """
+    Adds the value to each of the accumulator constituting the
+    CompoundAccumulator
```
[the remark this answers is not in the document]
Comment (preethiraghavan1) [reply pipeline_dp/accumulator.py]: Done

File: pipeline_dp/accumulator.py [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+    Args:
+      v: value to be added
```
[the remark this answers is not in the document]
Comment (preethiraghavan1) [reply pipeline_dp/accumulator.py]: Done

File: pipeline_dp/accumulator.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+    pass
+
+  @abc.abstractmethod
+  def compute_metrics(self):
+    pass
+
+
+class CompoundAccumulator(Accumulator):
+  """
+  Performs aggregations for a compund accumulators
```
[the remark this answers is not in the document]
Comment (preethiraghavan1) [reply pipeline_dp/accumulator.py]: Updated

File: pipeline_dp/accumulator.py (hunk 2 above)
Comment (preethiraghavan1) [reply pipeline_dp/accumulator.py]: Thanks for explaining! I misunderstood the first time!

File: tests/accumulator_test.py (hunk 3 above)
Comment (preethiraghavan1) [reply tests/accumulator_test.py:60]: :)

File: pipeline_dp/accumulator.py [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+  """
+
+  def __init__(self, accumulators: typing.Iterable['Accumulator'] = None):
+    self.accumulators = []
+    if accumulators:
+      # flatten the accumulators if the input is a CompundAccumulator
+      self.accumulators = [accumulator_expanded for accumulator in accumulators
```
[the remark this answers is not in the document]
Comment (preethiraghavan1) [reply pipeline_dp/accumulator.py]: got it!  Done!

File: pipeline_dp/accumulator.py [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+  @classmethod
+  def deserialize(cls, serialized_obj: str):
+    deserialized_obj = pickle.loads(serialized_obj)
+    if not isinstance(deserialized_obj, cls):
+      raise TypeError("The deserialized object is not of the right type")
+    return deserialized_obj
+
+class CompoundAccumulator(Accumulator):
+  """
+    Accumulator for computing multiple metrics.
```
Comment (dvadym) [inline pipeline_dp/accumulator.py:66]: Nit:
"""Accumulator ...

File: pipeline_dp/pipeline_operations.py [hunk 11]
```diff
@@ -1,7 +1,6 @@
 """Adapters for working with pipeline frameworks."""
 
 import random
-import collections
```
Comment (dvadym) [inline pipeline_dp/pipeline_operations.py:4]: Probably it's an incorrect merge in git. Please return back

**Review by dvadym — COMMENTED** [review]
Thanks! It looks great! I left some minor comments

File: pipeline_dp/accumulator.py [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+
+
+class Accumulator(abc.ABC):
+  """
+    Base class for all accumulators.
+    Accumulators are objects that encapsulate aggregations and computations of
+    differential private metrics.
+  """
+
+  @classmethod
```
Comment (dvadym) [inline pipeline_dp/accumulator.py:13]: During the reviewing [PR](https://github.com/OpenMined/PipelineDP/pull/49) I've realized that it's more convenient to have merge as a standalone function.

Could you please make merge a standalone function?

File: pipeline_dp/__init__.py [hunk 13]
```diff
@@ -6,3 +6,5 @@
 from pipeline_dp.pipeline_operations import LocalPipelineOperations
 from pipeline_dp.pipeline_operations import BeamOperations
 from pipeline_dp.pipeline_operations import SparkRDDOperations
+from pipeline_dp.accumulator import Accumulator
```
[the remark this answers is not in the document]
Comment (preethiraghavan1) [reply pipeline_dp/__init__.py]: Makes sense! Removed it

File: pipeline_dp/accumulator.py (hunk 12 above)
Comment (preethiraghavan1) [reply pipeline_dp/accumulator.py]: Made it a standalone function. What do you think of making it staticmethod in Accumulator?

File: pipeline_dp/accumulator.py [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+
+    Returns: self.
+    """
+    pass
+
+  @abc.abstractmethod
+  def add_accumulator(self, accumulator: 'Accumulator') -> 'Accumulator':
+    """
+      Merges the accumulator.
+      The difference between this and the merge function is that here it
```
[the remark this answers is not in the document]
Comment (preethiraghavan1) [reply pipeline_dp/accumulator.py]: Removed the line.

File: pipeline_dp/accumulator.py (hunk 10 above)
Comment (preethiraghavan1) [reply pipeline_dp/accumulator.py]: DOne

File: pipeline_dp/accumulator.py [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+    self.accumulators = accumulators
+
+  def add_value(self,  value):
+    for accumulator in self.accumulators:
+      accumulator.add_value(value)
+    return self
+
+  def add_accumulator(self, accumulator: 'CompoundAccumulator') -> \
+          'CompoundAccumulator':
+    # merges the accumulators of the CompoundAccumulators.
```
[the remark this answers is not in the document]
Comment (preethiraghavan1) [reply pipeline_dp/accumulator.py]: Done

File: pipeline_dp/accumulator.py [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+    # the expectation is that the input accumulators are of the same type and
+    # are in the same order.
+
+    if (len(accumulator.accumulators) != len(self.accumulators)
```
[the remark this answers is not in the document]
Comment (preethiraghavan1) [reply pipeline_dp/accumulator.py]: Let me know if the change looks good.

File: pipeline_dp/accumulator.py [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+        "Accumulators in the input are not of the same size "
+        + "or don't match the type/order of the base accumulators.")
+
+    for (base_accumulator, to_add_accumulator) in zip(self.accumulators,
+                                                      accumulator.accumulators):
+      base_accumulator.add_accumulator(to_add_accumulator)
+    return self
+
+  def compute_metrics(self):
+    # Computes the metrics for individual accumulator
```
[the remark this answers is not in the document]
Comment (preethiraghavan1) [reply pipeline_dp/accumulator.py]: Done

File: pipeline_dp/pipeline_operations.py (hunk 11 above)
Comment (preethiraghavan1) [reply pipeline_dp/pipeline_operations.py:4]: Done.

File: tests/accumulator_test.py [hunk 18]
```diff
@@ -0,0 +1,146 @@
...
+    self.sum_squares = np.sum([concat_acc.sum_squares
+                               for concat_acc in
+                               accumulators]) if accumulators else 0
+
+  def add_value(self, v):
+    self.sum_squares += v * v
+    return self
+
+  def add_accumulator(self,
+                      accumulator: 'MeanAccumulator') -> 'MeanAccumulator':
```
[the remark this answers is not in the document]
Comment (preethiraghavan1) [reply tests/accumulator_test.py]: Done.

File: pipeline_dp/accumulator.py [hunk 19]
```diff
@@ -0,0 +1,117 @@
...
+    return deserialized_obj
+
+
+class CompoundAccumulator(Accumulator):
+  """Accumulator for computing multiple metrics.
+
+    CompoundAccumulator contains one or more accumulators of other types for
+    computing multiple metrics.
+    For example it can contain [CountAccumulator,  SumAccumulator].
+    CompoundAccumulator delegates all operation to the internal accumulators.
```
Comment (dvadym) [inline pipeline_dp/accumulator.py:74]: Nit: operation**s**

File: pipeline_dp/accumulator.py [hunk 20]
```diff
@@ -0,0 +1,117 @@
...
+  Returns: Accumulator instance with merged values.
+  """
+  unique_accumulator_types = {type(accumulator).__name__ for accumulator in
+                              accumulators}
+  if len(unique_accumulator_types) > 1:
+    raise TypeError(
+      "Accumulators should all be of the same type. Found accumulators of "
+      + f"different types: ({','.join(unique_accumulator_types)}).")
+
+  return reduce(lambda acc1, acc2: acc1.add_accumulator(acc2), accumulators)
```
Comment (dvadym) [inline pipeline_dp/accumulator.py:23]: Nice implementation!

File: pipeline_dp/accumulator.py [hunk 21]
```diff
@@ -0,0 +1,117 @@
...
+def merge(accumulators: typing.Iterable[
+  'Accumulator'] = None) -> 'Accumulator':
+  """Merges the accumulators.
+
+  Args:
+    accumulators:
+
+  Returns: Accumulator instance with merged values.
+  """
+  unique_accumulator_types = {type(accumulator).__name__ for accumulator in
```
Comment (dvadym) [inline pipeline_dp/accumulator.py:16]: Let's assume that `add_accumulator` checks types:
1. The correctness of `add_accumulator` depends on whether types the same and 
2. `add_accumulator`  might be called without `merge`

Nit: remove types checks

File: pipeline_dp/accumulator.py [hunk 22]
```diff
@@ -0,0 +1,117 @@
...
+import typing
+import pickle
+from functools import reduce
+
+
+def merge(accumulators: typing.Iterable[
+  'Accumulator'] = None) -> 'Accumulator':
+  """Merges the accumulators.
+
+  Args:
```
Comment (dvadym) [inline pipeline_dp/accumulator.py:11]: Nit: the short description is enoughm remove starting from "Args:...

File: pipeline_dp/accumulator.py [hunk 23]
```diff
@@ -0,0 +1,117 @@
+import abc
+import typing
+import pickle
+from functools import reduce
+
+
+def merge(accumulators: typing.Iterable[
+  'Accumulator'] = None) -> 'Accumulator':
```
Comment (dvadym) [inline pipeline_dp/accumulator.py:8]: Nit: remove = None

File: pipeline_dp/accumulator.py [hunk 24]
```diff
@@ -0,0 +1,117 @@
...
+
+  def add_value(self, value):
+    for accumulator in self.accumulators:
+      accumulator.add_value(value)
+    return self
+
+  def add_accumulator(self, accumulator: 'CompoundAccumulator') -> \
+    'CompoundAccumulator':
+    """Merges the accumulators of the CompoundAccumulators.
+    The expectation is that the input accumulators are of the same type and
```
Comment (dvadym) [inline pipeline_dp/accumulator.py:88]: Nit: input -> internal

File: pipeline_dp/accumulator.py [hunk 25]
```diff
@@ -0,0 +1,117 @@
...
+    self.accumulators = accumulators
+
+  def add_value(self, value):
+    for accumulator in self.accumulators:
+      accumulator.add_value(value)
+    return self
+
+  def add_accumulator(self, accumulator: 'CompoundAccumulator') -> \
+    'CompoundAccumulator':
+    """Merges the accumulators of the CompoundAccumulators.
```
Comment (dvadym) [inline pipeline_dp/accumulator.py:87]: Nit: add a blank line after this line

File: pipeline_dp/accumulator.py [hunk 26]
```diff
@@ -0,0 +1,117 @@
...
+    The expectation is that the input accumulators are of the same type and
+    are in the same order."""
+
+    if len(accumulator.accumulators) != len(self.accumulators):
+      raise ValueError(
+        "Accumulators in the input are not of the same size."
+        + f" Expected size = {len(self.accumulators)}"
+        + f" received size = {len(accumulator.accumulators)}.")
+
+    expected_type_order = ",".join([type(accumulator).__name__ for
```
Comment (dvadym) [inline pipeline_dp/accumulator.py:97]: Nit: Checking of types might be done in slighly shorter and more efficient way (no need to create lists):

```
for acc1, acc2 in zip(..., ...):
  if type(acc1) is not type(acc2):
     raise ....        # here we can just provide the type of the 1st mismatched accumulators 
```

File: tests/accumulator_test.py [hunk 27]
```diff
@@ -0,0 +1,189 @@
...
+
+    with self.assertRaises(TypeError) as context:
+      base_compound_accumulator.add_accumulator(to_add_compound_accumulator)
+    self.assertEqual("""Accumulators in the input don't match the type/order of the base accumulators. 
+        Expected MeanAccumulator,SumOfSquaresAccumulator
+        received SumOfSquaresAccumulator,MeanAccumulator""",
+                     str(context.exception))
+
+  def test_adding_mismatched_accumulator_length_raises_exception(self):
+    mean_acc1 = MeanAccumulator().add_value(11)
```
Comment (dvadym) [inline tests/accumulator_test.py:66]: Thanks, I like thorough testing!

File: pipeline_dp/accumulator.py [hunk 28]
```diff
@@ -0,0 +1,117 @@
...
+    Args:
+      value: value to be added.
+
+    Returns: self.
+    """
+    pass
+
+  @abc.abstractmethod
+  def add_accumulator(self, accumulator: 'Accumulator') -> 'Accumulator':
+    """Merges the accumulator to self and returns self.
```
Comment (dvadym) [inline pipeline_dp/accumulator.py:45]: Please add comment:

That sub-class implementation is responsible for checking that types of self and accumulator are the same.

File: pipeline_dp/__init__.py [hunk 29]
```diff
@@ -6,3 +6,4 @@
 from pipeline_dp.pipeline_operations import LocalPipelineOperations
 from pipeline_dp.pipeline_operations import BeamOperations
 from pipeline_dp.pipeline_operations import SparkRDDOperations
+from pipeline_dp.accumulator import CompoundAccumulator
```
Comment (dvadym) [inline pipeline_dp/__init__.py:9]: Nit: remove from init.py

Explanation: CompoundAccumulator is not a part of Public API

**Review by dvadym — COMMENTED** [review]
Thanks it looks great! Just some minor comments

File: pipeline_dp/__init__.py (hunk 29 above)
Comment (preethiraghavan1) [reply pipeline_dp/__init__.py]: Sorry missed that.

File: pipeline_dp/accumulator.py (hunk 23 above)
Comment (preethiraghavan1) [reply pipeline_dp/accumulator.py]: Done

File: pipeline_dp/accumulator.py (hunk 22 above)
Comment (preethiraghavan1) [reply pipeline_dp/accumulator.py]: Done

File: pipeline_dp/accumulator.py (hunk 21 above)
Comment (preethiraghavan1) [reply pipeline_dp/accumulator.py]: Done

File: pipeline_dp/accumulator.py (hunk 20 above)
Comment (preethiraghavan1) [reply pipeline_dp/accumulator.py:9]: Thanks! :)

File: pipeline_dp/accumulator.py (hunk 28 above)
Comment (preethiraghavan1) [reply pipeline_dp/accumulator.py:31]: Added

File: pipeline_dp/accumulator.py (hunk 19 above)
Comment (preethiraghavan1) [reply pipeline_dp/accumulator.py]: Done

File: pipeline_dp/accumulator.py (hunk 25 above)
Comment (preethiraghavan1) [reply pipeline_dp/accumulator.py:76]: Done

File: pipeline_dp/accumulator.py (hunk 24 above)
Comment (preethiraghavan1) [reply pipeline_dp/accumulator.py]: Done

File: pipeline_dp/accumulator.py (hunk 26 above)
Comment (preethiraghavan1) [reply pipeline_dp/accumulator.py]: Got it! Let me know if this looks good!

File: tests/accumulator_test.py (hunk 27 above)
Comment (preethiraghavan1) [reply tests/accumulator_test.py:65]: Thanks for your thorough review!

**Review by dvadym — APPROVED** [review]
Thanks for implementing Accumulator framework!

**Pull request merged**